Convert "Add order note" screen to MVVM

hichamboushaba opened the pull request:
This is the first part of #2892, it refactors the "add order note" screen to MVVM.
As the PR is big enough, I'll do the "Add shipment Tracking fragment" refactoring in a second PR.

Please note that there is a small change in the flow, when adding a note, we will show a progress dialog: see p1607072347401500-slack-C6H8C3G23

To test: 
test different scenarios of adding notes, success, offline, error.


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
AddOrderNote MVVM, WIP

hichamboushaba committed:
Add a ViewModel and wire it to the fragment

hichamboushaba committed:
Add discard changes dialog and publishing loading dialog

hichamboushaba committed:
Remove the duplicate call to add note

hichamboushaba committed:
Remove unused classes

hichamboushaba committed:
Hide the add button if the note's text is empty

hichamboushaba committed:
Add AddOrderNoteViewModelTest

hichamboushaba committed:
Update order details tests

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt:
```diff
@@ -364,7 +365,8 @@ class IssueRefundViewModel @AssistedInject constructor(
                             if (reason.isNotBlank()) {
-                                noteRepository.createOrderNote(order.identifier, reason, false)
+                                val note = OrderNote(note = reason, isCustomerNote = false)
+                                orderDetailRepository.addOrderNote(order.identifier, order.remoteId, note)
```
This is a suspending function, which means we won't show the success refund message until the notes addition is done or fails, which is different from the previous behavior.
Though, I'm ignoring the result, to keep the same visible behavior as before.

hichamboushaba committed:
Fix ktlint issues

hichamboushaba committed:
Update tests to allow updating mocks before initializing the viewmodel

hichamboushaba committed:
Remove the duplicate add note function

AmandaRiu reviewed (changes requested):
Nice work @hichamboushaba. The code looks good, but I noticed an issue during testing. The "Add" button will disappear when rotating the device or restoring the app from background while the "add note" screen is open. Typing doesn't make it visible so the only fix is to cancel out of the note and start over.

hichamboushaba commented:
Thanks @AmandaRiu for the test, I will check what's wrong.

hichamboushaba committed:
Restore the state of add button

hichamboushaba committed:
Merge branch 'develop' into issue/2892-mvvm-order-cont

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/notes/AddOrderNoteFragment.kt

hichamboushaba committed:
Convert AddOrderNoteFragment to use the fragment ctor

hichamboushaba commented:
@AmandaRiu it appears this bug is not related to my changes, as I can reproduce on the latest beta as well, I created an issue to track it #3296, I tried investigating it a bit, but couldn't figure out what's going on yet, my guess is that it's related to the toolbar changes.
I added the commit 78c3130d751b5ced86f31f3ab1cb758116063331 to make sure that the state of the add button is restored, so I'm pretty sure when we fix the other but, we won't have any issue with my changes.

I also merged the last develop changes to use the new BaseFragment constructor, and converted the binding to a local parameter: d1ce930668c192dc637f8f64dcfd6fd24a73aa6a

hichamboushaba committed:
Fix a typo on the test

AmandaRiu reviewed (approved):
LGTM! :shipit:
